In [10]:
%load_ext autoreload
%autoreload 2
import sys
sys.path.append('../src/')

In [11]:
from modelling.sgd import SGD_optimizer
from data.create_dataset import *
from visualization.visualize import *
from sklearn.model_selection import  train_test_split
from sklearn.preprocessing import StandardScaler

In [13]:
from sklearn.datasets import load_breast_cancer
from sklearn.linear_model import LogisticRegression
cancer = load_breast_cancer()

In [5]:
cancer.data.shape

(569, 30)

In [106]:
cancer.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

Ok so data looks fairly simple, it's a design matrix with 40 features and 569 samples. We simply input it to the model like normal. So the inputs are continuous, but the target data, i.e the z data are 0's and 1's. So we're esentially creating a linear function that produces a result which is continuous and is then fed into sigmoid, which then classifies, or simply gives probability.

In [181]:
X_train, X_test, z_train, z_test = train_test_split(cancer.data,cancer.target)
z_train.shape

(426,)

In [182]:
X_train, X_test, z_train, z_test = train_test_split(cancer.data,cancer.target, test_size=0.2)
Xscaler = StandardScaler().fit(X_train)
#zscaler = StandardScaler().fit(z_train)

X_train = Xscaler.transform(X_train)
#z_train = zscaler.transform(z_train)


In [197]:
from modelling.nn import NeuralNetwork
nn = NeuralNetwork(hidden_activation = 'relu',output_activation = 'sigmoid',
                   momentum = 0.5,w_init='glorot',batch_size = 1,n_epochs = 1)
nn.fit(X_train,z_train)
#nn.predict(cancer.data)

In [198]:
nn.predict(X_train)

array([[9.92287577e-01],
       [4.16516395e-04],
       [2.51369544e-01],
       [7.87325239e-01],
       [9.99210268e-01],
       [8.88091681e-01],
       [9.98364110e-01],
       [9.43764861e-01],
       [9.94195507e-01],
       [3.23581313e-03],
       [7.59171516e-01],
       [9.96906121e-01],
       [9.83487000e-01],
       [9.90850934e-01],
       [9.98810680e-01],
       [9.99243749e-01],
       [1.47272453e-02],
       [9.92994701e-01],
       [7.42776003e-04],
       [9.50162280e-04],
       [1.82197203e-02],
       [8.28247959e-01],
       [9.93335886e-01],
       [9.99471287e-01],
       [1.26622604e-02],
       [9.99535703e-01],
       [8.03965196e-01],
       [9.98368595e-01],
       [7.54133265e-05],
       [3.06497967e-01],
       [9.92720060e-01],
       [9.97902824e-01],
       [9.75756173e-03],
       [9.97201589e-01],
       [7.60958581e-05],
       [9.32211636e-01],
       [7.86745889e-02],
       [3.33168239e-03],
       [1.38959100e-04],
       [9.67215671e-01],


So overflow when just feeding this directly into my network. What about keras or sklearn?

In [199]:
from tensorflow.keras import initializers
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

#initializer = tf.keras.initializers.RandomNormal(mean=0., stddev=1.)
initializer = tf.keras.initializers.GlorotUniform()
init_b = tf.keras.initializers.Constant(value=0.001)


model = keras.Sequential()

model.add(layers.Dense(100, kernel_initializer=initializer,bias_initializer= init_b, input_shape=(30,)))
model.add(layers.Activation('relu'))

model.add(layers.Dense(50, kernel_initializer=initializer,bias_initializer= init_b))
model.add(layers.Activation('relu'))

model.add(layers.Dense(1, kernel_initializer=initializer,bias_initializer= init_b))
model.add(layers.Activation('sigmoid'))

opt = keras.optimizers.SGD(learning_rate=0.01,momentum=0.5)
model.compile(loss='binary_crossentropy', optimizer=opt)

In [200]:
model.fit(X_train,z_train,batch_size = 32, epochs=1000,verbose=0)

In [201]:
model.predict(cancer.data)

array([[0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],

In [163]:
model.get_layer(index=2).get_weights()[1]

array([-0.2512364 ,  0.27123645], dtype=float32)

In [87]:
model.summary()

Model: "sequential_11"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_22 (Dense)             (None, 100)               3100      
_________________________________________________________________
activation_22 (Activation)   (None, 100)               0         
_________________________________________________________________
dense_23 (Dense)             (None, 100)               10100     
_________________________________________________________________
activation_23 (Activation)   (None, 100)               0         
_________________________________________________________________
dense_24 (Dense)             (None, 1)                 101       
_________________________________________________________________
activation_24 (Activation)   (None, 1)                 0         
Total params: 13,301
Trainable params: 13,301
Non-trainable params: 0
_________________________________________________

In [186]:
from sklearn.neural_network import MLPClassifier

nn_sk = MLPClassifier(max_iter = 1000, hidden_layer_sizes=(100,50),
                     activation = 'relu',solver='sgd',
                     learning_rate_init=0.01,batch_size = 32,momentum= 0.5,alpha=0.0,tol=0)
nn_sk.fit(X_train,z_train)

c:\users\oliver\appdata\local\programs\python\python39\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:614: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


MLPClassifier(alpha=0.0, batch_size=32, hidden_layer_sizes=(100, 50),
              learning_rate_init=0.01, max_iter=1000, momentum=0.5,
              solver='sgd', tol=0)

In [188]:
#print(nn_sk.predict_proba(X_train))
print('\n')
print(nn_sk.coefs_[-1].shape)
print(nn_sk.classes_)
print(nn_sk.get_params())
nn_sk.predict(X_train)



(50, 1)
[0 1]
{'activation': 'relu', 'alpha': 0.0, 'batch_size': 32, 'beta_1': 0.9, 'beta_2': 0.999, 'early_stopping': False, 'epsilon': 1e-08, 'hidden_layer_sizes': (100, 50), 'learning_rate': 'constant', 'learning_rate_init': 0.01, 'max_fun': 15000, 'max_iter': 1000, 'momentum': 0.5, 'n_iter_no_change': 10, 'nesterovs_momentum': True, 'power_t': 0.5, 'random_state': None, 'shuffle': True, 'solver': 'sgd', 'tol': 0, 'validation_fraction': 0.1, 'verbose': False, 'warm_start': False}


array([1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1,
       1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1,
       1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1,
       0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1,
       1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1,
       0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0,
       1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0,
       1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0,
       0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0,
       1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0,
       1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1,

In [123]:
from tensorflow.keras.utils import to_categorical
from sklearn.preprocessing import OneHotEncoder
y = to_categorical(cancer.target)
y2 = OneHotEncoder().fit_transform(cancer.target.reshape(-1,1))
print(y.shape)
print(y2.shape)

(569, 2)
(569, 2)


In [118]:
cancer.target.shape

(569,)

In [126]:
len(cancer.target.shape)

1

In [149]:
from sklearn.preprocessing import LabelBinarizer
lb = LabelBinarizer()
y3 = lb.fit_transform(cancer.target)
y3[0]

array([0])

So what might be the case is simply that sklearn and keras implements methods to avoid numerical instabilities so that computation can continue. With  inf values here and there an activation with sigmoid will simply result in 1, or minus inf will result in 0. So perhaps somehow the probability outputted, not being 1 btw, has something to do either with the distribution of the data set, or that it magically somehow converges to 0.62 something. Perhaps something to do with e?

So the weights of sklearn and keras are not infinite, they are something. But perhaps those are the weights that reached the limit of maximum number in terms of computational stability?

Either way, turned out that standardizing the data helped majorly in avoiding exploding gradient.

In [210]:
from modelling.logreg import SGD_logreg

logreg = SGD_logreg()
logreg.fit(X_train,z_train.reshape(-1,1))

In [211]:
logreg.predict(X_train)

array([[1],
       [0],
       [0],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [0],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [0],
       [1],
       [0],
       [0],
       [0],
       [1],
       [1],
       [1],
       [0],
       [1],
       [1],
       [1],
       [0],
       [0],
       [1],
       [1],
       [0],
       [1],
       [0],
       [1],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [1],
       [0],
       [0],
       [1],
       [1],
       [0],
       [1],
       [0],
       [0],
       [1],
       [1],
       [1],
       [1],
       [0],
       [1],
       [1],
       [1],
       [0],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [1],
       [1],
       [1],
       [1],
       [0],
       [1],
       [1],
       [1],
       [1],
       [0],
    

In [213]:
logreg.score(X_train,z_train.reshape(-1,1))

0.967032967032967

In [218]:
from sklearn.linear_model import LogisticRegression
logreg_sk = LogisticRegression(fit_intercept = False, solver='lbfgs',)
logreg_sk.fit(X_train,z_train)

LogisticRegression(fit_intercept=False)

In [220]:
logreg_sk.score(X_test,z_test)

0.40350877192982454